### ISSUE TITLE
Concept: Decomposition and Integration

### ISSUE THREAD

**Issue by smetsth** [dwmkerr/hacker-laws#12]
Part of this is based on Tesler's law on complexity where you write the following:
"This complexity can be moved, but not eliminated". To me this forms the base of decomposing systems into smaller parts. Let's call those smaller parts modules for the sake of this discussion.

I don't want to focus on decomposing systems into modules here, but I do want to point out two ideas I find important:
- [On the Criteria To Be Used in Decomposing Systems into Modules
D.L. Parnas](http://sunnyday.mit.edu/16.355/parnas-criteria.html)
- [Structured Design by Yourdon and Constantine](https://www.win.tue.nl/~wstomv/quotes/structured-design.html#6)

Instead I want to focus on the integration part. Since it's still a system that's being developed, most of the modules being developed will depend on other modules. An order processing module might for example depend on a warehousing module.

Is there any 'law' that describes this integration? Or preferably a 'law' that combines both of these?
I once had a discussion with a colleage where we discussed the cost per module versus the cost of integration. More visually it's represented in the image below.
- If you only have 1 module, there is no cost to integrate
- If you have a lot of modules, the cost of integration is high
- In the middle there is a balance where each module is just the right size, and there are just the right amount of modules.
The exact number will ofcourse vary depending on the size of the system.
![image](https://user-images.githubusercontent.com/1089629/46582494-03805c00-ca48-11e8-8bff-5046c3202912.png)

Have you come across books or articles that describe that cost of integration? Or even better discuss both at the same time? It would be great to have a better understanding about these topics.

**Comment by dwmkerr** [comment]
Hi @smetsth,

I'd like to find a way to include this concept. Essentially it also covers one of the most fundamental challenges of software engineering, which is 'how do we design components?'. Picking where to draw boundaries between components, and how to interface them is one of the core challenges for any kind of engineering. Whether it is breaking one large function into many small ones (making easy functions, but more complexity to handle passing parameters around), or breaking a vast enterprise monolith into many small isolated services.

I _suppose_ Domain Driven Design (#13) would touch on this, but I've not yet found a good single 'principle' which covers it. I like the visual you've done which presents this - I could see this making sense on an enterprise level, but even in the internals of a single system (i.e. smaller modules in a mobile app - each module is easier to manage, but the system as a whole has more touchpoints).

If I can't find a specific example soon, I'll probably create a 'User Contributed' section and add there. Have you found anything else on this since raising this issue?

Cheers!

**Comment by smetsth** [comment]
I totally agree with you that this can be applied from the smallest parts of an application to the to the integration of systems even. I haven't found a specific example yet, even though I've been looking for it for quite some time.

Domain driven development certainly addresses integration of code. Articles and papers on coupling and cohesion do as well. But I haven't found a catchy phrase that explains the break even point in the picture.

**Comment by dwmkerr** [comment]
In that case I'm going to leave this issue open and let it stew for a bit. I've got a backlog of other easier laws to introduce so I'll work through them, and then add to this as I discover more!
